From 172ce9d4282e408fe60b699c432390b9f6e3f74a Mon Sep 17 00:00:00 2001
From: Ronald Wampler <rdwampler@gmail.com>
Date: Wed, 22 Apr 2015 21:47:53 -0400
Subject: [PATCH] Replace `thread_local` with boost equivalent

The new version of clang (Apple LLVM version 6.1.0) as shipped with Xcode 6.3, disabled TLS. According to http://clang.llvm.org/cxx_status.html, in order to support `thread_local`, the C++ runtime library from g++-4.8 or later is needed.

For now, we can use the boost `thread_specific_ptr`. This is probably a reasonable solution since 1) it should be portable with old and future versions of (Apple's) clang and 2) requires no additional dependencies.
---
 Frameworks/crash/src/info.cc | 8 ++++++--
 target                       | 2 ++
 2 files changed, 8 insertions(+), 2 deletions(-)

diff --git ./Frameworks/crash/src/info.cc ./Frameworks/crash/src/info.cc
index 5c98180..b83789b 100644
--- ./Frameworks/crash/src/info.cc
+++ ./Frameworks/crash/src/info.cc
@@ -1,4 +1,5 @@
 #include "info.h"
+#include <boost/thread/tss.hpp>
 #include <oak/debug.h>
 
 /* CrashReporter info */
@@ -58,8 +59,11 @@ namespace
 
 	static stack_t& stack ()
 	{
-		thread_local stack_t stack;
-		return stack;
+		static boost::thread_specific_ptr<stack_t> stackPtr;
+		if(!stackPtr.get())
+			stackPtr.reset(new stack_t);
+
+		return *stackPtr;
 	}
 }
 
diff --git ./target ./target
index 5d8e2.7..6731511 100644
--- ./target
+++ ./target
@@ -28,6 +28,8 @@ LN_FLAGS += -rpath @executable_path/../Frameworks
 CXX_FLAGS += -I"$capnp_prefix/include"
 LN_FLAGS  += -L"$capnp_prefix/lib"
 
+LN_FLAGS += -lboost_thread-mt
+
 PRELUDE = Shared/PCH/prelude.*
 
 TARGETS  = vendor/*/target
